{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Running cisTopic with input features from ScreSeg-fi on multiome ATAC modality\n",
    "\n",
    "For dimensionality reduction and clustering of scATAC-seq data we found that using Wolfgang's ScregSeg model and cisTopic gave us the best clustering resolution. I want to try this with the multiome-ATAC modality.\n",
    "\n",
    "I made a few adaptions of the standard 10X scATAC tutorial to get this to work with the ATAC modality of multiome data: http://htmlpreview.github.io/?https://github.com/aertslab/cisTopic/blob/master/vignettes/10X_workflow.html"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Set up installation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set project head directory\n",
    "project_path<-\"/fish/\"\n",
    "# Specify sample name\n",
    "sample<-\"multiome-fish\"\n",
    "# extract metadata\n",
    "metadata<-read.csv(paste(project_path,\"sample_metadata.csv\", sep=\"\"))\n",
    "#metadata\n",
    "sample_metadata<-metadata[metadata$Sample_Name==sample,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# already installed devtools::install_github(\"aertslab/cisTopic\") with envrionment\n",
    "library(cisTopic)\n",
    "library(GenomicRanges)\n",
    "lapply(c('Rsubread', 'umap', 'Rtsne', 'ComplexHeatmap', \n",
    "         'fastcluster', 'data.table', 'rGREAT', 'ChIPseeker', \n",
    "         'TxDb.Hsapiens.UCSC.hg19.knownGene', 'org.Hs.eg.db', \n",
    "         'densityClust'),\n",
    "       require, character.only = TRUE)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Input data\n",
    "From cisTopic manual\n",
    "Starting from the CellRanger ATAC fragments file and a bed file with potential rgeulatory regions. For more information regards the fragments file, please visit: https://support.10xgenomics.com/single-cell-atac/software/pipelines/latest/output/fragments. \n",
    "\n",
    "I will actually use the informative states generated by ScregSeg and exported as a bed file 'informative_12_rarest_states.bed'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The scregseg output bed file has several columns in addition to chr, start, end. These need to be removed for compatibility with cistopic so I removed them with the following awk command."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the cellranger peak bed files have only chr, start and end so I need to parse this \n",
    "#system(\"awk -F'\\t' '{print $1\"\\t\"$2\"\\t\"$3}' informative_12_rarest_states.bed > informative_12_rarest_states.3col.bed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To create a cisTopic obeject some metrics information that is output from cellranger is used. The way this is used is available on the github page [https://github.com/aertslab/cisTopic/blob/04cecbb9d1112fcc1a6edc28b5a506bcb49f2803/R/InitializecisTopic.R](https://github.com/aertslab/cisTopic/blob/04cecbb9d1112fcc1a6edc28b5a506bcb49f2803/R/InitializecisTopic.R) This serves two purposes:\n",
    "1. Selection of barcodes called as cells\n",
    "2. Additional metrics in the cisTopic object\n",
    "\n",
    "However the metrics file output by cellranger-arc (for multiome) has a slightly different format because it includes information on both modalities. So I selected the colums that match the columns found in the scATAC metrics files, and renamed these to match. NB a few columns were missing, these seem to mainly be addition metrics e.g. blacklist or DNase sensitive regions so I hope they will just be optional downstream."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in metrics file from multiome\n",
    "metrics_preparse<-read.csv(metrics)\n",
    "\n",
    "#select columns from that match the metrics from scATAC\n",
    "metrics_parsed<-metrics_preparse[c('barcode', 'atac_raw_reads','atac_dup_reads','atac_chimeric_reads',\n",
    "                                   'atac_unmapped_reads','atac_lowmapq','atac_mitochondrial_reads', \n",
    "                                   #no \"passed_filters\" or \"cell_id\"\n",
    "                                   'is_cell', 'atac_TSS_fragments', \n",
    "                                   # no DNase_sensitive_region_fragments, no enhancer_region_fragments, no promoter_region_fragments\n",
    "                                   # no on_target_fragments, blacklist_region_fragments\n",
    "                                   'atac_peak_region_fragments','atac_peak_region_cutsites')\n",
    "                                 ]\n",
    "colnames(metrics_parsed)<-c('barcode','total','duplicate','chimeric',\n",
    "                            'unmapped','lowmapq','mitochondrial',\n",
    "                            #'passed_filters','cell_id',\n",
    "                            'is__cell_barcode','TSS_fragments',\n",
    "                            #'DNase_sensitive_region_fragments','enhancer_region_fragments','promoter_region_fragments',\n",
    "                            #'on_target_fragments','blacklist_region_fragments',\n",
    "                            'peak_region_fragments','peak_region_cutsites')\n",
    "metrics_parsed_path<-paste0(pathTo10X, '/per_barcode_metrics_parsed.csv')\n",
    "write.csv(metrics_parsed, \n",
    "         file=metrics_parsed_path, quote=FALSE, row.names=FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now set up the paths for the cisTopic object as in the tutorial but using 'per_barcode_metrics_parsed.csv' instead of 'per_barcode_metrics.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pathTo10X <- paste0(project_path,sub(\"./\", \"\", sample_metadata$cellranger_count_out_path))\n",
    "fragments <- paste0(pathTo10X, '/atac_fragments.tsv.gz')\n",
    "regions <- 'informative_12_rarest_states.3col.bed'\n",
    "#metrics <- paste0(pathTo10X, '/per_barcode_metrics.csv')\n",
    "metrics_parsed_path<-paste0(pathTo10X, '/per_barcode_metrics_parsed.csv')\n",
    "# check whether this will work because format of metrics might be different for multiome\n",
    "cisTopicObject <- createcisTopicObjectFrom10X(fragments, regions, metrics_parsed_path, project.name=sample)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This works and the dimensionality reduction can be performed as in the tutorial."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R [conda env:cistopicanalysis] *",
   "language": "R",
   "name": "conda-env-cistopicanalysis-r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
